registry/reference.go

```diff
@@ -56,33 +56,51 @@ type Reference struct {
 	Reference string
 }
 
-// ParseReference parses a string into a artifact reference.
-// If the reference contains both the tag and the digest, the tag will be
-// dropped.
+// ParseReference parses a string into an artifact reference.
+// If the reference contains both the tag and the digest, the tag will be dropped.
 // Digest is recognized only if the corresponding algorithm is available.
-func ParseReference(raw string) (Reference, error) {
-	parts := strings.SplitN(raw, "/", 2)
+func ParseReference(image string) (Reference, error) {
+	parts := strings.SplitN(image, "/", 2)
 	if len(parts) == 1 {
+		// Invalid Form
 		return Reference{}, fmt.Errorf("%w: missing repository", errdef.ErrInvalidReference)
 	}
 	registry, path := parts[0], parts[1]
+	/*
+		The parameter `image` can thus be parsed (lowercase implies opaque, uppercase implies resolute)
+			<---------------- image ---------------------------------------------------> | Decode `image`
+			<=== REGISTRY ===> / <------------------ path -----------------------------> |  - Decode `path``
+			<=== REGISTRY ===> / <=== REPOSITORY ===> * <---------- reference ---------> |    - Decode `reference`
+			<=== REGISTRY ===> / <=== REPOSITORY ===> @ <=================== DIGEST ===> |      - Valid Form A
+			<=== REGISTRY ===> / <=== REPOSITORY ===> : <=== TAG ===> @ <=== DIGEST ===> |      - Valid Form B
+			<=== REGISTRY ===> / <=== REPOSITORY ===> : <=== TAG ======================> |      - Valid Form C
+			<=== REGISTRY ===> / <=== REPOSITORY ======================================> |    - Valid Form D
+
+		Note: DIGEST is itself, if present, of the form
+			<=========== DIGEST ==========>
+			<=== ALGO ===> : <=== HASH ===>
+
+		Note: Furthermore, in the case of Valid Form B, the TAG is dropped without any validation or consideration
+			  whatsoever.
+	*/
```

Comment:
can we add unit tests for some of these cases instead of a comment?

Reply:
I think the comment still helps, this polymorphic token deserves a docstring to show all the possible ways it can be used.  Not sure if the placement of this doc is the best though, maybe it can move somewhere else that's more appropriate.

Reply:
Maybe I should rephrase this... can we add unit tests for this instead of *just* a comment?

Reply:
Given that this function hasn't changed in any meaningful way, only annotated, why do you think it needs tests as part of this change?  I don't think it's a bad idea, just trying to see what made you see that as a blocker.

I'll take a look at adding tests meanwhile, maybe it will be quick.

Reply:
Let me know if you think differently, but all 4 valid forms, as well as the invalid forms are already being tested.  I've annotated them in a change which I'll post in a bit.

Reply:
[Done](https://github.com/oras-project/oras-go/pull/237/commits/150d0c325d5913eb96237d35bb5bc76fa5c6268f)

Reply:
Will now look at where best to put that docstring.

Reply:
[Done](https://github.com/oras-project/oras-go/pull/237/commits/833528e0ec96ee3c8909361cc85263bf40aaade2)